File: lua/problem_40.lua
local function singleNumber(nums)
    local result = 0
    for bit = 0, 31 do
        local total = 0
        for _, n in ipairs(nums) do
            total = total + (math.floor(n / 2^bit) % 2)
        end
        if total % 3 ~= 0 then result = result + 2^bit end
    end
    return result
end

local function problem40()
    print("Problem 40:")
    assert(singleNumber({6,1,3,3,3,6,6}) == 1,  "assertion failed")
    assert(singleNumber({13,19,13,13})   == 19, "assertion failed")
    assert(singleNumber({0,0,0,5})       == 5,  "assertion failed")
    print("All Tests Passed")
end

return { problem40 = problem40 }

File: c/problem_40.c
#include "problem_40.h"
#include <stdio.h>
#include <assert.h>

long long singleNumber40(const int *nums, int n) {
    long long result = 0;
    for (int bit = 0; bit < 32; bit++) {
        long long total = 0;
        for (int i = 0; i < n; i++) total += (nums[i] >> bit) & 1;
        if (total % 3 != 0) result |= (1LL << bit);
    }
    return result;
}

void problem40(void) {
    printf("Problem 40:\n");
    int a[] = {6,1,3,3,3,6,6};
    int b[] = {13,19,13,13};
    int c[] = {0,0,0,5};
    assert(singleNumber40(a, 7) == 1);
    assert(singleNumber40(b, 4) == 19);
    assert(singleNumber40(c, 4) == 5);
    printf("All Tests Passed\n");
}
